pgmpy/inference/CausalInference.py

```diff
@@ -0,0 +1,291 @@
+            A set of variables Z satisfies the back-door criterion relative to an ordered pair ofvariabies (Xi, Xj) in a DAG G if:
+                (i) no node in Z is a descendant of Xi; and
+                (ii) Z blocks every path between Xi and Xj that contains an arrow into Xi.
+
+        TODO:
+          * Backdoors are great, but the most general things we could implement would be Ilya Shpitser's ID and
+            IDC algorithms. See [his Ph.D. thesis for a full explanation]
+            (https://ftp.cs.ucla.edu/pub/stat_ser/shpitser-thesis.pdf). After doing a little reading it is clear
+            that we do not need to immediatly implement this.  However, in order for us to truly account for
+            unobserved variables, we will need not only these algorithms, but a more general implementation of a DAG.
+            Most DAGs do not allow for bidirected edges, but it is an important piece of notation which Pearl and
+            Shpitser use to denote graphs with latent variables.
+
+        Parameters
+        ----------
+        X: str
+            Intervention Variable
+        Y: str
+            Target Variable
+        """
+        try:
+            assert X in self.observed_variables
+            assert Y in self.observed_variables
+        except AssertionError:
+            raise AssertionError("Make sure both X and Y are observed.")
+
+        if self.is_valid_backdoor_adjustment_set(X, Y, Z=frozenset()):
+            return frozenset()
+
+        possible_adjustment_variables = (
+            set(self.observed_variables)
+            - {X} - {Y}
+            - set(nx.descendants(self.dag, X))
+        )
+
+        valid_adjustment_sets = []
+        for s in _powerset(possible_adjustment_variables):
+            super_of_complete = any([
+                vs.intersection(set(s)) == vs
+                for vs in valid_adjustment_sets
+            ])
+            if super_of_complete:
+                continue
+            if self.is_valid_backdoor_adjustment_set(X, Y, s):
+                valid_adjustment_sets.append(frozenset(s))
+
+        return frozenset(valid_adjustment_sets)
+
+    def is_valid_frontdoor_adjustment_set(self, X, Y, Z=None):
+        """
+        Test whether Z is a valid frontdoor adjustment set for estimating the causal impact of X on Y via the frontdoor
+        adjustment formula.
+
+        Parameters
+        ----------
+        X: str
+            Intervention Variable
+        Y: str
+            Target Variable
+        Z: set
+            Adjustment variables
+        """
+        Z = _variable_or_iterable_to_set(Z)
+
+        # 0. Get all directed paths from X to Y.  Don't check further if there aren't any.
+        directed_paths = list(nx.all_simple_paths(self.dag, X, Y))
+
+        if directed_paths == []:
+            return False
+
+        # 1. Z intercepts all directed paths from X to Y
+        unblocked_directed_paths = [
```

Comment:
Would really suggest writing a for loop in these cases. Makes the code much more readable.